### PR TITLE
Instrument writes to topic partitions

### PR DESCRIPTION
**Pull request by dasch**
Subscribers can track writes to individual partitions. This allows monitoring throughput per topic and partition.

### CODE REVIEW

**Comment by dasch** [comment]
@bquorning I could use some feedback on this.

File: lib/kafka/produce_operation.rb [hunk 1]
```diff
@@ -1,4 +1,25 @@
 module Kafka
+  # A produce operation attempts to send all messages in a buffer to the Kafka cluster.
+  # Since topics and partitions are spread among all brokers in a cluster, this usually
+  # involves sending requests to several or all of the brokers.
+  #
+  # ## Instrumentation
+  #
+  # When executing the operation, a `append_message_set.kafka` notification will be
```
Comment (bquorning) [inline lib/kafka/produce_operation.rb]: `s/a/an/`

File: lib/kafka/produce_operation.rb [hunk 2]
```diff
@@ -1,4 +1,25 @@
...
+  # emitted for each message set that was successfully appended to a topic partition.
+  # The following keys will be found in the payload:
+  #
+  # * `:topic` — the topic that was written to.
+  # * `:partition` — the partition that the message set was appended to.
+  # * `:offset` — the offset assigned to the first message in the message set appended
```
Comment (bquorning) [inline lib/kafka/produce_operation.rb]: “assigned” sounds wrong here. Could it be “the offset of the first message…”?

File: lib/kafka/produce_operation.rb [hunk 3]
```diff
@@ -1,4 +1,25 @@
...
+  #   to the partition.
+  # * `:message_count` — the number of messages that were appended.
+  #
+  # If there was an error appending the message set the key `:exception` will be set
```
Comment (bquorning) [inline lib/kafka/produce_operation.rb]: Comma after “set”, please.

File: lib/kafka/produce_operation.rb (hunk 2 above)
Comment (dasch) [inline lib/kafka/produce_operation.rb]: It's actually taken from the Kafka protocol docs. I can change it, though.

File: lib/kafka/produce_operation.rb [hunk 4]
```diff
@@ -1,4 +1,25 @@
 module Kafka
+  # A produce operation attempts to send all messages in a buffer to the Kafka cluster.
+  # Since topics and partitions are spread among all brokers in a cluster, this usually
+  # involves sending requests to several or all of the brokers.
+  #
+  # ## Instrumentation
+  #
+  # When executing the operation, a `append_message_set.kafka` notification will be
+  # emitted for each message set that was successfully appended to a topic partition.
```
Comment (bquorning) [inline lib/kafka/produce_operation.rb]: All the time I’m reading “message set” as if a message has been set. Would it be possible to write `message_set` instead?

Comment (dasch) [inline lib/kafka/produce_operation.rb]: Nah, it's a set of messages. Nothing peculiar about that. It's also the standard Kafka nomenclature, and I don't want to diverge from that too much.

File: lib/kafka/produce_operation.rb (hunk 2 above)
Comment (bquorning) [inline lib/kafka/produce_operation.rb]: You know its meaning better than I do. Only change it if you agree it’s better.

File: lib/kafka/produce_operation.rb (hunk 4 above)
Comment (bquorning) [inline lib/kafka/produce_operation.rb]: Sure, ok.

**Comment by bquorning** [comment]
:+1:

**Pull request merged**